src/java/org/apache/cassandra/index/sai/disk/vector/OnDiskOrdinalsMap.java

```diff
@@ -227,21 +229,59 @@ public void close()
         }
     }
 
+    /**
+     * not thread safe
+     */
     private class FileReadingOrdinalsView implements OrdinalsView
     {
         RandomAccessReader reader = fh.createReader();
         private final long high = (segmentEnd - 8 - rowOrdinalOffset) / 8;
+        private int lastFoundRowId = -1;
+        private long lastFoundRowIdIndex = -1;
 
         /**
          * @return order if given row id is found; otherwise return -1
          */
         @Override
         public int getOrdinalForRowId(int rowId) throws IOException
         {
+            if (rowId <= lastFoundRowId)
+                logger.warn("repeated read for rowId {}, last found rowid: {}", rowId, lastFoundRowId);
```

Comment:
I think we want to track the last passed `rowId` and then we can throw an exception if it is `lastRowId <= rowId`.

Because the `lastFoundRowId` can skip rows if the row id does not have an associated vector, it is possible `rowId <= lastFoundRowId`.